### PR TITLE
Generic repository decorator

### PR DESCRIPTION
**Pull request by 19majkel94**
This PR unify the `@OrmRepository` decorator to work with `Repository`, `TreeRepository`, `MongoRepository` and custom repositories.

### CODE REVIEW

**Comment by 19majkel94** [comment]
@pleerock Honestly, I'm afraid of this changes 😆 
I've done only the type check if this decorator usage compiles.
However we have no test suite so the new complicated decorator has to be debugged by you 😜

File: src/decorators/OrmRepository.ts [hunk 1]
```diff
@@ -1,21 +1,179 @@
...
+                    `Repository, MongoRepository, TreeRepository or custom repository class type.`
+                );
+            }
+            repositoryType = paramTypes[index];
+        }
+        // if the parameter has been aplied to class property
+        else {
+            const propertyType: Function | undefined = Reflect.getOwnMetadata("design:type", object, propertyName);
+            if (!propertyType) {
+                throw new Error(
```
Comment (pleerock) [inline src/decorators/OrmRepository.ts]: maybe better to extract errors in to "error" classes? It makes code so huge and hardly readable

File: src/decorators/OrmRepository.ts [hunk 2]
```diff
@@ -1,21 +1,179 @@
...
+                    `Cannot get reflected type for a property "${propertyName}" of ${object.constructor.name} class. ` +
+                    `Make sure you have turned on an "emitDecoratorMetadata": true, option in tsconfig.json ` +
+                    `and that you have imported "reflect-metadata" on top of the main entry file in your application.` +
+                    `And make sure that you have annotated the property type correctly with: ` +
+                    `Repository, MongoRepository, TreeRepository or custom repository class type.`
+                );
+            }
+            repositoryType = propertyType;
```
Comment (pleerock) [inline src/decorators/OrmRepository.ts]: If Im not wrong `connection.getRepository` always returns you `MongoRepository` for mongodb, `TreeRepository` for tree entities, and regular `Repository` for regular entities, so I think you don't really need "repository type" to make different calls to connection.

**Comment by pleerock** [comment]
I'll keep it open until 0.1.0 final released. Once I release 0.1.0 I'll check all those libraries as well by myself.

**Comment by pleerock** [comment]
or you can provide a test cases and infrastructure 😆

File: src/decorators/OrmRepository.ts (hunk 2 above)
Comment (19majkel94) [reply src/decorators/OrmRepository.ts]: If you are sure, I can do this, but the same switch-case already exist in `@TransactionRepository` decorator code 😉

File: src/decorators/OrmRepository.ts (hunk 1 above)
Comment (19majkel94) [reply src/decorators/OrmRepository.ts]: Right, I haven't thought that the decorator code will grow up to this amount of lines 😉

**Comment by 19majkel94** [comment]
> I'll keep it open until 0.1.0 final released. Once I release 0.1.0 I'll check all those libraries as well by myself.

Maybe let's merge it do the `next` branch and release on npm as `0.1.0-beta.0` pointed by `typeorm-typedi-extensions@next` in npm. So I could verify if it works ok in my real world project because

> or you can provide a test cases and infrastructure

I don't know how to configure the test environment for database-things, like you have docker and other stuffs in TypeORM 😉

**Comment by pleerock** [comment]
> Maybe let's merge it do the next branch and release on npm as 0.1.0-beta.0 pointed by typeorm-typedi-extensions@next in npm. 

too much bureaucracy for such a small package 😆 

> So I could verify if it works ok in my real world project because

you can do `gulp package` just like in any other library and get contents of build/package into your node_modules renamed to package name

**Comment by 19majkel94** [comment]
> you can do gulp package just like in any other library and get contents of build/package into your node_modules renamed to package name

Right, I can even do `npm link`! I'm so stupid after 10h of coding in work 😜 
If it works ok, I will release it on npm as `@19majkel94/typeorm-typedi-extensions` 😉

**Comment by pleerock** [comment]
I'll merge it, but will release it a bit later (once 0.1.0 is finished)

**Comment by 19majkel94** [comment]
> (once 0.1.0 is finished)
I guess that will be this month, right? Because now github readme points to new version but npm install download the old, incompatible version 😞

**Comment by pleerock** [comment]
yes Im going to release it next week

**Pull request merged**